libs/datamodel/core/src/transform/ast_to_dml/db/walkers.rs

```diff
@@ -357,3 +389,118 @@ impl<'ast, 'db> PrimaryKeyWalker<'ast, 'db> {
         self.attribute.name
     }
 }
+
+#[derive(Copy, Clone)]
+pub(crate) struct ExplicitRelationWalker<'ast, 'db> {
+    pub(crate) field_a: RelationFieldWalker<'ast, 'db>,
+    pub(crate) field_b: RelationFieldWalker<'ast, 'db>,
+    #[allow(dead_code)]
+    pub(super) relation: &'db Relation<'ast>,
+    pub(crate) db: &'db ParserDatabase<'ast>,
+}
+
+impl<'ast, 'db> ExplicitRelationWalker<'ast, 'db> {
+    pub(super) fn referencing_model(&self) -> ModelWalker<'ast, 'db> {
+        self.field_a.model()
+    }
+
+    pub(super) fn referenced_model(&self) -> ModelWalker<'ast, 'db> {
+        self.field_b.model()
+    }
+
+    pub(super) fn referencing_field(&self) -> RelationFieldWalker<'ast, 'db> {
+        self.field_a
+    }
+
+    /// The scalar fields defining the relation on the referenced model.
+    pub(super) fn referenced_fields(&'db self) -> impl ExactSizeIterator<Item = ScalarFieldWalker<'ast, 'db>> + 'db {
+        let f = move |field_id: &ast::FieldId| {
+            let model_id = self.referenced_model().model_id;
+
+            ScalarFieldWalker {
+                model_id,
+                field_id: *field_id,
+                db: self.db,
+                scalar_field: &self.db.types.scalar_fields[&(model_id, *field_id)],
+            }
+        };
+
+        match self.referencing_field().relation_field.references.as_ref() {
+            Some(references) => references.iter().map(f),
+            None => [].iter().map(f),
+        }
+    }
+
+    /// The scalar fields on the defining the relation on the referencing model.
+    pub(super) fn referencing_fields(&'db self) -> impl ExactSizeIterator<Item = ScalarFieldWalker<'ast, 'db>> + 'db {
+        let f = move |field_id: &ast::FieldId| {
+            let model_id = self.referencing_model().model_id;
+
+            ScalarFieldWalker {
+                model_id,
+                field_id: *field_id,
+                db: self.db,
+                scalar_field: &self.db.types.scalar_fields[&(model_id, *field_id)],
+            }
+        };
+
+        match self.referencing_field().relation_field.fields.as_ref() {
```

Comment:
I think this is from the AST, so it will sometimes be empty. I think it's fine for now, but we should be mindful to not expose that in the future public API.

Reply:
We should definitely guarantee this is a valid datamodel before making the API public.